usr/man/man1/ipcs.1
.TH ipcs 1
.\" Last modified by BAM on 8-Aug-85  1300  
.\"
.SH NAME
ipcs \- report inter-process communication facilities status
.SH SYNTAX
.ft B
ipcs [
.IR options ]
.SH DESCRIPTION
The
.Pn ipcs
command provides information about active,
inter-process communication facilities,
message queues, shared memory, and semaphores
that are currently active in the system.
The information is displayed in columns and
is controlled by the following
.IR options :
.IP -q
Print information about active message queues.
.IP -m
Print information about active shared memory segments.
.IP -s
Print information about active semaphores.
.LP
If any of the options
.RB -q,
.RB -m,
or
.RB -s
are specified, information about only those indicated will be printed.
If none of these three are specified, information about all three will
be printed.
.IP -b
Print biggest allowable size information.  Maximum number of bytes in
messages on queue for message queues, size of segments for shared
memory, and number of semaphores in each set for semaphores.  See below
for meaning of columns in a listing.
.IP -c
Print creator's login name and group name.  
.IP -o
Print information on outstanding usages.  (Number of messages on queue
and total number of bytes in messages on queue for message queues and
number of processes attached to shared memory segments.)
.IP -p
Print process number information. Process ID of last process to send a
message and process ID of last process to receive a message on message
queues and process ID of creating process and process ID of last process
to attach or detach on shared memory segments. 
.IP -t
Print time information.  (Time of the last control operation that
changed the access permissions for all facilities.  Time of last
.PN msgsnd
and last
.PN msgrcv
on message queues, last
.PN shmat
and last
.PN shmdt
on shared memory, last semop(2) on semaphores.) 
.IP -a
Use all print
.IR options .
(This is a shorthand notation for
.B -b, -c, -o, -p
and
.BR -t .)
.IP -C corefile
Use the file
.I corefile
in place of /dev/kmem.
.IP -N namelist
The argument will be taken as the name of an alternate
.I namelist
.RB (/vmunix
is the default).
.LP
The column headings and the meaning of the columns in an
.PN ipcs
listing are given below; the letters in parentheses indicate the
.I options
that cause the corresponding heading to appear;
.B all
means that the heading always appears.  Note that these
.I options
only determine what information is provided for each facility; they do
.I not
determine which facilities will be listed.
.TP 20
T (all)	
Type of facility:
.RS
.nf
q	message queue;
m	shared memory segment;
s	semaphore.
.fi
.RE
.TP 20
ID (all)	
The identifier for the facility entry.
.TP 20
KEY (all)
The key used as an argument to
.PN msgget ,
.PN semget
or
.PN shmget
to create the facility entry.  Note:  The key of a shared memory
segment is changed to IPC_PRIVATE
when the segment has been removed until all processes attached to the
segment detach it.
.TP 20
MODE (all)
The facility access modes and flags: The mode consists of 11 characters
that are interpreted as follows:
.nf
The first two characters are:
.fi
.RS
.RS
.TP
R
if the process is waiting on a
.PN msgrcv;
.TP
S
if a process is waiting on a
.PN msgsnd;
.TP
D
if the associated shared memory segment has been removed.  It will
disappear when the last process attached to the segment detaches it;
.TP
C
if the associated shared memory segment is to be clear when the first
attach is executed;
.TP
-
if the corresponding special flag is not set.
.IP
The next 9 characters are interpreted as three sets of three bits each.
The first set refers to the owner's permissions; the next, to
permissions of others in the user-group of the facility entry; and the
last to all others.  Within each set, the first character indicates
permission to write or alter the facility entry, and the last character
is currently unused.
.PP
.nf
The permissions are indicated as follows:
r	if read permission is granted;
w	if write permission is granted;
a	if alter permission is granted;
.fi
-   if the indicated permission is
.I not
granted;
.RE
.RE
.TP 20
OWNER (all)  
The login name of the owner of the facility entry.
.TP 20
GROUP (all)  
The group name of the group of the owner of the facility entry.
.TP 20
CREATOR (a,c)
The login name of the creator of the facility entry.
.TP 20
CGROUP (a,c)
The group name of the group of the creator of the facility entry.
.TP 20
CBYTES (a,o)
The number of bytes in messages currently outstanding on the associated
message queue.
.TP 20
QNUM (a,o)
The number of messages currently
outstanding on the associated message queue.
.TP 20
QBYTES (a,b)
The maximum number of bytes allowed in messages outstanding
on the associated message queue.
.TP 20
LSPID (a,p)
The process ID of the last process to send a message to the associated
queue.
.TP 20
LRPID (a,p)
The process ID of the last process
to receive a message from the associated
queue.
.TP 20
STIME (a,t)
The time the last message was sent to the associated queue.
.TP 20
RTIME (a,t)
The time the last message was received from the associated queue.
.TP 20
CTIME (a,t)
The time the associated entry was created or changed.
.TP 20
NATTCH (a,o)
The number of processes attached to the associated shared memory
segment.
.TP 20
SEGSZ (a,b)
The size of the associated shared memory segment.
.TP 20
CPID (a,p)
The process ID of the creator of the shared memory entry.
.TP 20
LPID (a,p)
The process ID of the last process to attach or detach the shared memory
segment.
.TP 20
ATIME (a,t)
The time the last attach was completed to the associated shared memory
segment.
.TP 20
DTIME (a,t)
The time the last detach was completed on the associated shared memory
segment.
.TP 20
NSEMS (a,b)
The number of semaphores in the set associated with the semaphore
entry.
.TP 20
OTIME (a,t)
The time the last semaphore operation was completed on the set
associated with the semaphore entry.
.SH RESTRICTIONS
Things can change while
.PN ipcs
is running.  The picture it gives is only a close approximation to
reality.
.SH FILES
.nf
/vmunix		system namelist
/dev/kmem	memory
/etc/passwd	user names
/etc/group	group names
.fi
.SH SEE ALSO
msgop(2), semop(2), shmop(2).


usr.bin/ipcs.c


#ifndef lint
static	char	*sccsid = "@(#)ipcs.c	1.2	(ULTRIX)	1/29/87";
#endif lint

/************************************************************************
 *									*
 *			Copyright (c) 1984 by				*
 *		Digital Equipment Corporation, Maynard, MA		*
 *			All rights reserved.				*
 *									*
 *   This software is furnished under a license and may be used and	*
 *   copied  only  in accordance with the terms of such license and	*
 *   with the  inclusion  of  the  above  copyright  notice.   This	*
 *   software  or  any  other copies thereof may not be provided or	*
 *   otherwise made available to any other person.  No title to and	*
 *   ownership of the software is hereby transferred.			*
 *									*
 *   This software is  derived  from  software  received  from  the	*
 *   University    of   California,   Berkeley,   and   from   Bell	*
 *   Laboratories.  Use, duplication, or disclosure is  subject  to	*
 *   restrictions  under  license  agreements  with  University  of	*
 *   California and with AT&T.						*
 *									*
 *   The information in this software is subject to change  without	*
 *   notice  and should not be construed as a commitment by Digital	*
 *   Equipment Corporation.						*
 *									*
 *   Digital assumes no responsibility for the use  or  reliability	*
 *   of its software on equipment which is not supplied by Digital.	*
 *									*
 ************************************************************************/
/*
 *
 *   Modification history:
 *
 *  30 Apr 85 -- depp
 *	New file to perform system maintenance on System V IPC
 *
 */

/*
**	ipcs - IPC status
**	Examine and print certain things about message queues, semaphores,
**		and shared memory.
*/

#include	<machine/pte.h>

#include	<sys/param.h>
#include	<sys/vm.h>
#include	<sys/ipc.h>
#include	<sys/msg.h>
#include	<sys/sem.h>
#include	<sys/shm.h>
#include	<nlist.h>
#include	<fcntl.h>
#include	<sys/time.h>
#include	<grp.h>
#include	<pwd.h>
#include	<stdio.h>

#define	TIME	0
#define	MSG	1
#define	SEM	2
#define	SHM	3
#define	MSGINFO	4
#define	SEMINFO	5
#define	SHMINFO	6

struct nlist    nl[] =
{			/* name list entries for IPC facilities */
    { "_time" },{ "_msgque" },{ "_sema" },{ "_smem" },{ "_msginfo" },
    { "_seminfo" }, { "_sminfo" }, 
    { "_Sysmap" },
#define SSYSMAP		7
    { "_Syssize" },
#define SSYSSIZE	8
    { NULL },
};


char    chdr[] = "T     ID     KEY        MODE       OWNER    GROUP",
 /* common header format */
        chdr2[] = "  CREATOR   CGROUP",
 /* c option header format */
       *name = "/vmunix",	/* name list file */
       *mem = "/dev/kmem",	/* memory file */
        opts[] = "abcmopqstC:N:";/* allowable options for getopt */
extern char *optarg;		/* arg pointer for getopt */
int     bflg,		/* biggest size: segsz on m; qbytes on q; nsems on s */
        cflg,		/* creator's login and group names */
        mflg,		/* shared memory status */
        oflg,		/* outstanding data: nattch on m; cbytes, qnum on q */
        pflg,		/* process id's: lrpid, lspid on q; cpid, lpid on m */
        qflg,		/* message queue status */
        sflg,		/* semaphore status */
        tflg,		/* times: atime, ctime, dtime on m; ctime,
				   rtime, stime on q; ctime, otime on s */
	nflg,		/* indicator that a particular IPC capability is
				   currently not being used */
        err;		/* option error count */

extern int  optind;		/* option index for getopt */

extern char *ctime ();
extern struct group *getgrgid ();
extern struct passwd   *getpwuid ();
extern long lseek ();

char	*kmemf, *memf, *swapf, *nlistf;
int	kmem, kflg, swap = -1;
#define clear(x) 	((int)x & 0x7fffffff)

main (argc, argv)
int     argc;			/* arg count */
char  **argv;			/* arg vector */
{
    register int    i,		/* loop control */
                    md,		/* memory file file descriptor */
                    o;		/* option flag */
    time_t time;		/* date in memory file */
    struct shmid_ds mds;	/* shared memory data structure */
    struct shminfo  shminfo;	/* shared memory information structure */
    struct msqid_ds qds;	/* message queue data structure */
    struct msginfo  msginfo;	/* message information structure */
    struct semid_ds sds;	/* semaphore data structure */
    struct seminfo  seminfo;	/* semaphore information structure */

    /* Go through the options and set flags. */
    while ((o = getopt (argc, argv, opts)) != EOF)
	switch (o)
	{
	    case 'a': 
		bflg = cflg = oflg = pflg = tflg = 1;
		break;
	    case 'b': 
		bflg = 1;
		break;
	    case 'c': 
		cflg = 1;
		break;
	    case 'C': 
		mem = optarg;
		break;
	    case 'm': 
		mflg = 1;
		break;
	    case 'N': 
		name = optarg;
		break;
	    case 'o': 
		oflg = 1;
		break;
	    case 'p': 
		pflg = 1;
		break;
	    case 'q': 
		qflg = 1;
		break;
	    case 's': 
		sflg = 1;
		break;
	    case 't': 
		tflg = 1;
		break;
	    case '?': 
		err++;
		break;
	}
    if (err || (optind < argc))
    {
	fprintf (stderr,
		"usage:  ipcs [-abcmopqst] [-C corefile] [-N namelist]\n");
	exit (1);
    }
    if ((mflg + qflg + sflg) == 0)
	mflg = qflg = sflg = 1;

    /* Check out namelist and memory files. */
    nlist (name, nl);
    if (!nl[TIME].n_value)
    {
	fprintf (stderr, "ipcs:  no namelist\n");
	exit (1);
    }
	openfiles();
/*
    if ((md = open (mem, O_RDONLY)) < 0)
    {
	fprintf (stderr, "ipcs:  no memory file\n");
	exit (1);
    }
*/
    klseek (kmem, (long) nl[TIME].n_value, 0);

/*
    puts ("before reade time");
    printf ("&time = %o\n", &time);
*/
    reade (kmem, &time, sizeof (time));
/*
    puts ("after reade time");
*/

    printf ("\nIPC status from %s as of %s", mem, ctime (&time));

    /* Print Message Queue status report. */
    if (qflg)
    {
	if (nl[MSG].n_value)
	{
	    i = 0;
	    klseek (kmem, (long) nl[MSGINFO].n_value, 0);
	    reade (kmem, &msginfo, sizeof (msginfo));
	    klseek (kmem, (long) nl[MSG].n_value, 0);
	    printf ("Message Queues:\n%s%s%s%s%s%s\n", chdr,
		    cflg ? chdr2 : "",
		    oflg ? " CBYTES  QNUM" : "",
		    bflg ? " QBYTES" : "",
		    pflg ? " LSPID LRPID" : "",
		    tflg ? "   STIME    RTIME    CTIME " : "");
	} else
	{
	    i = msginfo.msgmni;
	    printf ("Message Queue facility not in system.\n\n");
	}
	nflg = 1;
	while (i < msginfo.msgmni)
	{
	    reade (kmem, &qds, sizeof (qds));
	    if (!(qds.msg_perm.mode & IPC_ALLOC))
	    {
		i++;
		continue;
	    }
	    hp ('q', "SRrw-rw-rw-", &qds.msg_perm, i++, msginfo.msgmni);
	    if (oflg)
		printf ("%7u%6u", qds.msg_cbytes, qds.msg_qnum);
	    if (bflg)
		printf ("%7u", qds.msg_qbytes);
	    if (pflg)
		printf ("%6u%6u", qds.msg_lspid, qds.msg_lrpid);
	    if (tflg)
	    {
		tp (qds.msg_stime);
		tp (qds.msg_rtime);
		tp (qds.msg_ctime);
	    }
	    printf ("\n");
	    nflg = 0;
	}
	if (nflg)
		printf("*** No message queues are currently defined ***\n\n");
	else
		printf("\n");
    }

    /* Print Shared Memory status report. */
    if (mflg)
    {
	if (nl[SHM].n_value)
	{
	    i = 0;
	    klseek (kmem, (long) nl[SHMINFO].n_value, 0);
	    reade (kmem, &shminfo, sizeof (shminfo));
	    klseek (kmem, (long) nl[SHM].n_value, 0);
	    printf ("Shared Memory\n%s%s%s%s%s%s\n", chdr,
		cflg ? chdr2 : "",
		oflg ? " NATTCH" : "",
		bflg ? "  SEGSZ" : "",
		pflg ? "  CPID  LPID" : "",
		tflg ? "   ATIME    DTIME    CTIME " : "");
	}
	else
	{
	    i = shminfo.shmmni;
	    printf ("Shared Memory facility not in system.\n\n");
	}
	nflg = 1;
	while (i < shminfo.shmmni)
	{
	    reade (kmem, &mds, sizeof (mds));
	    if (!(mds.shm_perm.mode & IPC_ALLOC))
	    {
		i++;
		continue;
	    }
	    hp ('m', "DCrw-rw-rw-", &mds.shm_perm, i++, shminfo.shmmni);
	    if (oflg)
		printf ("%7u", mds.shm_nattch);
	    if (bflg)
		printf ("%7d", mds.shm_segsz);
	    if (pflg)
		printf ("%6u%6u", mds.shm_cpid, mds.shm_lpid);
	    if (tflg)
	    {
		tp (mds.shm_atime);
		tp (mds.shm_dtime);
		tp (mds.shm_ctime);
	    }
	    printf ("\n");
	    nflg = 0;
	}
	if (nflg)
		printf("*** No shared memory segments are currently defined ***\n\n");
	else
		printf("\n");
    }

    /* Print Semaphore facility status. */
    if (sflg)
    {
	if (nl[SEM].n_value)
	{
	    i = 0;
	    klseek (kmem, (long) nl[SEMINFO].n_value, 0);
	    reade (kmem, &seminfo, sizeof (seminfo));
	    klseek (kmem, (long) nl[SEM].n_value, 0);
	    printf ("Semaphores\n%s%s%s%s\n", chdr,
		cflg ? chdr2 : "",
		bflg ? " NSEMS" : "",
		tflg ? "   OTIME    CTIME " : "");
	}
	else
	{
	    i = seminfo.semmni;
	    printf ("Semaphore facility not in system.\n\n");
	}
	nflg = 1;
	while (i < seminfo.semmni)
	{
	    reade (kmem, &sds, sizeof (sds));
	    if (!(sds.sem_perm.mode & IPC_ALLOC))
	    {
		i++;
		continue;
	    }
	    hp ('s', "--ra-ra-ra-", &sds.sem_perm, i++, seminfo.semmni);
	    if (bflg)
		printf ("%6u", sds.sem_nsems);
	    if (tflg)
	    {
		tp (sds.sem_otime);
		tp (sds.sem_ctime);
	    }
	    printf ("\n");
	    nflg = 0;
	}
	if (nflg)
		printf("*** No semaphores are currently defined ***\n\n");
	else
		printf("\n");
    }
    exit (0);
}

struct pte *Sysmap = 0;

klseek(fd, loc, off)
	int fd;
	long loc;
	int off;
{
	static int	sizeSysmap;
int a;
extern errno;

	if( kflg && Sysmap == 0)
		{/* initialize Sysmap */

		sizeSysmap = nl[SSYSSIZE].n_value * sizeof( struct pte);
		Sysmap = (struct pte *)calloc( sizeSysmap, 1);
		a = lseek( kmem, (long)clear( nl[SSYSMAP].n_value), 0);
if(a == -1)
printf("sizeSysmap = %d    a = %d    errno = %d\n",
sizeSysmap, a, errno);
		if( (a = read( kmem, Sysmap, sizeSysmap)) != sizeSysmap)
			{
printf("sizeSysmap = %d    a = %d    errno = %d\n",
sizeSysmap, a, errno);
printf("kmem = %d    Sysmap = 0x%x\n", kmem, Sysmap);
			cantread( "system page table", kmemf);
			exit(1);
			}
		}
	if( kflg && (loc&0x80000000))
		{/* do mapping for kernel virtual addresses */
		struct pte *ptep;

		loc &= 0x7fffffff;
		ptep = &Sysmap[btop(loc)];
		if( (char *)ptep - (char *)Sysmap > sizeSysmap)
			{
			printf( "no system pte for %s\n", loc);
			exit(1);
			}
		if( ptep->pg_v == 0)
			{
			printf( "system pte invalid for %x\n", loc);
			exit(1);
			}
		loc = (off_t)((loc&PGOFSET) + ptob(ptep->pg_pfnum));
		}
	(void) lseek(fd, (long)loc, off);
}

openfiles()
/*
openfiles(argc, argv)
	char **argv;
*/
{

	kmemf = "/dev/kmem";
/*
	if (kflg)
		kmemf = argc > 2 ? argv[2] : "/vmcore";
*/
	kmem = open(kmemf, 0);
	if (kmem < 0) {
		perror(kmemf);
		exit(1);
	}
/*
	if (kflg)  {
		mem = kmem;
		memf = kmemf;
	} else {
		memf = "/dev/mem";
		mem = open(memf, 0);
		if (mem < 0) {
			perror(memf);
			exit(1);
		}
	}
	if (kflg == 0 || argc > 3) {
		swapf = argc>3 ? argv[3]: "/dev/drum";
		swap = open(swapf, 0);
		if (swap < 0) {
			perror(swapf);
			exit(1);
		}
	}
*/
}

/*
**	reade - read with error exit
*/

reade (f, b, s)
int     f;			/* fd */
char   *b;			/* buffer address */
int     s;			/* size */
{
    if (read (f, b, s) != s)
    {
	perror ("ipcs:  read error");
	printf ("%o\n", b);
	exit (1);
    }
}

/*
**	hp - common header print
*/

hp (type, modesp, permp, slot, slots)
char    type,			/* facility type */
       *modesp;			/* ptr to mode replacement characters */
register struct ipc_perm   *permp;/* ptr to permission structure */
int     slot,			/* facility slot number */
        slots;			/* # of facility slots */
{
    register int    i,		/* loop control */
                    j;		/* loop control */
    register struct group  *g;	/* ptr to group group entry */
    register struct passwd *u;	/* ptr to user passwd entry */

/*
    printf ("%c%7d%s%#8.8x ", type, slot + slots * permp -> seq,
	    permp -> key ? " " : " 0x", permp -> key);
	    */
    printf ("%c%7d %10d ", type, slot + slots * permp -> seq, permp -> key);
    for (i = 02000; i; modesp++, i >>= 1)
	printf ("%c", (permp -> mode & i) ? *modesp : '-');
    if ((u = getpwuid (permp -> uid)) == NULL)
	printf ("%9d", permp -> uid);
    else
	printf ("%9.8s", u -> pw_name);
    if ((g = getgrgid (permp -> gid)) == NULL)
	printf ("%9d", permp -> gid);
    else
	printf ("%9.8s", g -> gr_name);
    if (cflg)
    {
	if ((u = getpwuid (permp -> cuid)) == NULL)
	    printf ("%9d", permp -> cuid);
	else
	    printf ("%9.8s", u -> pw_name);
	if ((g = getgrgid (permp -> cgid)) == NULL)
	    printf ("%9d", permp -> cgid);
	else
	    printf ("%9.8s", g -> gr_name);
    }
}

/*
**	tp - time entry printer
*/

tp (time)
time_t time;			/* time to be displayed */
{
    register struct tm *t;	/* ptr to converted time */

    if (time)
    {
	t = localtime (&time);
	printf (" %2d:%2.2d:%2.2d", t -> tm_hour, t -> tm_min, t -> tm_sec);
    }
    else
	printf (" no-entry");
}

cantread(what, fromwhat)
	char *what, *fromwhat;
{

	fprintf(stderr,
		"ipcs: error reading %s from %s\n", what, fromwhat);
}
